pkg/cloudmeta/metadata.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (C) 2024 ScyllaDB
+
+package cloudmeta
+
+import (
+	"context"
+	"errors"
+	"time"
+)
+
+// InstanceMetadata represents metadata returned by cloud provider.
+type InstanceMetadata struct {
+	InstanceType  string
+	CloudProvider CloudProvider
+}
+
+// CloudProvider is enum of supported cloud providers.
+type CloudProvider string
+
+// CloudProviderAWS represents aws provider.
+var CloudProviderAWS CloudProvider = "aws"
+
+// CloudMetadataProvider interface that each metadata provider should implement.
+type CloudMetadataProvider interface {
+	Metadata(ctx context.Context) (InstanceMetadata, error)
+}
+
+// CloudMeta is a wrapper around various cloud metadata providers.
+type CloudMeta struct {
+	providers []CloudMetadataProvider
+
+	providerTimeout time.Duration
+}
+
+// NewCloudMeta creates new CloudMeta provider.
+func NewCloudMeta() (*CloudMeta, error) {
+	const defaultTimeout = 5 * time.Second
+
+	awsMeta, err := NewAWSMetadata("")
+	if err != nil {
+		return nil, err
+	}
+
+	return &CloudMeta{
+		providers: []CloudMetadataProvider{
+			awsMeta,
+		},
+		providerTimeout: defaultTimeout,
+	}, nil
+}
+
+// GetInstanceMetadata tries to fetch instance metadata from AWS, GCP, Azure providers in order.
```

Comment:
Is the order important? 
In theory we could query the metadata in parallel - which is not needed for this use case, but why would we specify that?

Reply:
I think it's not important, but I decided for now to keep them in order of popularity, so we the most popular one is queried first and etc ... so basically I just specified what behavior to expect from this function. 

yes, we may want to change that later, but it depends on how this package will be used. 
There is also a possibility (theoretical) that we may want to enable/disable some providers based on configuration, e.g. query only aws or gcp. 

let me know if doesn't make sense for you